src/main/rx/ghst.c

```diff
@@ -0,0 +1,259 @@
+ * Len          u8      Length includes the packet ID, but not the CRC
+ * CRC          u8
+ * 
+ * Ghost packets are designed to be as short as possible, for minimum latency.
+ * 
+ * Note that the GHST protocol does not handle, itself, failsafe conditions. Packets are passed from 
+ * the Ghost receiver to Betaflight as and when they arrive. Betaflight itself is responsible for
+ * determining when a failsafe is necessary based on dropped packets. 
+ * 
+  */
+
+#define GHST_FRAME_LENGTH_ADDRESS       1
+#define GHST_FRAME_LENGTH_FRAMELENGTH   1
+#define GHST_FRAME_LENGTH_TYPE_CRC      1
+
+STATIC_UNIT_TESTED uint8_t ghstFrameCRC(void)
+{
+    // CRC includes type and payload
+    uint8_t crc = crc8_dvb_s2(0, ghstFrame.frame.type);
+    for (int i = 0; i < ghstFrame.frame.len - GHST_FRAME_LENGTH_TYPE_CRC - 1; ++i) {
+        crc = crc8_dvb_s2(crc, ghstFrame.frame.payload[i]);
+    }
+    return crc;
+}
+
+// Receive ISR callback, called back from serial port
+STATIC_UNIT_TESTED void ghstDataReceive(uint16_t c, void *data)
+{
+    UNUSED(data);
+
+    static uint8_t ghstFrameIdx = 0;
+    const timeUs_t currentTimeUs = microsISR();
+
+    if (cmpTimeUs(currentTimeUs, ghstFrameStartAtUs) > GHST_MAX_FRAME_TIME_US) {
+        // Character received after the max. frame time, assume that this is a new frame
+        ghstFrameIdx = 0;
+    }
+
+    if (ghstFrameIdx == 0) {
+        // timestamp the start of the frame, to allow us to detect frame sync issues
+        ghstFrameStartAtUs = currentTimeUs;
+    }
+
+    // assume frame is 5 bytes long until we have received the frame length
+    // full frame length includes the length of the address and framelength fields
+    const int fullFrameLength = ghstFrameIdx < 3 ? 5 : ghstFrame.frame.len + GHST_FRAME_LENGTH_ADDRESS + GHST_FRAME_LENGTH_FRAMELENGTH;
+
+    if (ghstFrameIdx < fullFrameLength) {
+        ghstFrame.bytes[ghstFrameIdx++] = (uint8_t)c;
+        if (ghstFrameIdx >= fullFrameLength) {                  // buffer is full?
+            ghstFrameIdx = 0;
+            const uint8_t crc = ghstFrameCRC();
+            if (crc == ghstFrame.bytes[fullFrameLength - 1]) {
+                switch (ghstFrame.frame.type)
+                {
+                    case GHST_UL_RC_CHANS_HS4_5TO8:
+                    case GHST_UL_RC_CHANS_HS4_9TO12:
+                    case GHST_UL_RC_CHANS_HS4_13TO16:
+                        if (ghstFrame.frame.addr == GHST_ADDR_FC) {
+                            lastRcFrameTimeUs = currentTimeUs;
+                            ghstFrameDone = true;
+                            memcpy(&ghstChannelDataFrame, &ghstFrame, sizeof(ghstFrame));
+                        }
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+        }
+    }
+}
+STATIC_UNIT_TESTED uint8_t ghstFrameStatus(rxRuntimeState_t *rxRuntimeState)
+{
+    UNUSED(rxRuntimeState);
+    int iStartIdx = 4;
+
+    if (ghstFrameDone) {
```

Comment:
This needs to be `volatile`, or else concurrent access between ISR and regular code will not work reliably.

Reply:
Done